### DISCUSSION TITLE
Convert.ToString should not return nullable string

### DISCUSSION THREAD

**Discussion by nathan130200** [dotnet/runtime#109285] [General]
As title, `Convert.ToString(object?)` and its overload `Convert.ToString(object?, IFormatProvider?)` should not return string with nullable annotation:

```cs
public static string? ToString(object? value, IFormatProvider? provider)
{
    if (value is IConvertible ic) // already check not null and IConvertible and .ToString does not return: string?
        return ic.ToString(provider);
    if (value is IFormattable formattable) // already check: not null and IFormattable and .ToString does not return: string?
        return formattable.ToString(null, provider);
    return value == null ? string.Empty : value.ToString(); // if value is null, just returns an empty string instead.
}
```

I know the `value.ToString()` may return null string but instead maybe it should be replaced with:

```cs
return value?.ToString() ?? string.Empty;
```

**Comment by tannergooding** [comment]
This would be a breaking change, would modify a behavior we've had for 20 years, would make it no longer possible to discern when the input was itself `null`, and would deviate from calling `value.ToString()` directly in the case `value` was `non-null`.

All of these are undesirable changes and would remove part of the purpose/intent of the API.

If you have a need for `null` output to be normalized, you should create a wrapper or utility function that handles that and use it throughout your codebase.

**Reply by nathan130200** [reply]
The nullable annotations feature in .NET is not even 20 years old.

Even VS/VSCode or any C# language service when overriding `Object.ToString` call on custom types does not add nullable annotation to the string (aka string?). It also does not make sense to override ToString method to return only null.

[SharpLab - Sample](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYBGAWAChGA3AQygAIMBPABxgCSeAM4ZWeMDA4BeDnhgB3DgFluAFT4wAFAEoA3Iwo0AnFp78ho8ZP2NDVFes0BvRh3ccKAZg4RmMKCgAS0IpIwAGDjUIAGUMYLwAc103D1cGD0zPAHYOYFZhGAA6aLiE5P0OAHoqjmEACwgcABtCDgA5AFEANU6AJTypPBbmjnqAmFT3AF9GaaA==)

But anyway... It was just an observation.

**Reply by tannergooding** [reply]
*Edit:* Minor fix to the behavior description below

> The nullable annotations feature in .NET is not even 20 years old.

The behavior of `Convert.ToString()` is 20 years old, it's been around since .NET Framework 1.0 and has always returned `string.Empty` when `value is null` and `null` when `value.ToString()` returns `null`. This is what would be breaking.

The annotations are simply surfacing, to the consumer of the API, the behavior that we've always had.

> It also does not make sense to override ToString method to return only null.

The consideration is that types and APIs do exist that return `null` for `ToString()`, thus `Convert.ToString(object)` may return `null`. They may not always return `null`, but they can return `null` and it would be a breaking change to force them to no longer do so.

**Reply by tannergooding** [reply]
> Only return null if value.ToString() (return value is null) returns null.

Which is a breaking change to modify in one of our oldest APIs. It's not something we're interested in doing.

There are trivial workarounds if you need a different behavior for your code base.

**Reply by nathan130200** [reply]
When `value is null` it returns `string.Empty` and not null. That's i'm pointing. Only return null if `value.ToString()` (return value is null) returns null.
